README.md
# ActionTestScript

ActionTestScript is a structured and readable testing language used to create reliable and performant GUI automated tests campaigns.
Tests scripts are defined by a sequence of 'actions' executed on web or desktop application.
Scripts written in ATS are converted into java classes and then executed using Selenium and TestNG frameworks.

Only following browsers are available for the moment with ATS : *Chrome, Edge, Firefox and Opera*.

With ATS, tests designers are only focused on the functional actions to do and don't have to worry about technical considerations.

## Getting Started

Here is a simple example of ATS script :

```
channel-start -> myFirstChannel -> chrome
goto-url -> google.com
keyboard -> automated testing$key(ENTER) -> INPUT [id = lst-ib]
click -> A [text = Test automation - Wikipedia]
scroll -> 300
scroll -> 0 -> A [font-style = italic, text = Graphical user interface testing]
over -> IMG [src =~ .*Test_Automation_Interface.png]
channel-close -> myFirstChannel
```

This script start Chrome browser, navigate to url 'google.com', enter 'automated test' string into input text field and hit 'enter' key.
The script click on the wikipedia link to open the page and do some mouse over and scroll actions.
At the end of the script the started channel called 'myFirstChannel' is closed and terminate the browser process.

Pretty simple no ?

### Prerequisites

* You have to install a standard Java 9 JDK into the folder of your choice (JRE 9 server distribution is working too).
* Apache Maven or Ant to generate and compile ATS project.
* Install TestNG plugin on your favorite java IDE (for Eclipse : http://testng.org/doc/eclipse.html).
* Read TestNG documentation about suite management.


### Installing

Download ATS components : http://www.actiontestscript.com/ats.zip .

You can unzip archive into the folder of your choice, but if you do not install ATS on *[User-Home-Directory]/.actiontestscript* folder, you have to create an environment variable named **ATS_HOME** and set it's value to your ATS installation folder.

After installation of ATS components you can create ATS projects with the following folder structure.

```
ATS Project folder
	/libs/
	/src/
		/assets/
			/data/
			/lang/
			/resources/
		/exec/
		/main/
			/ats/
				/subscripts/
			/java/
	/target/
	/.atsProjectProperties
	/pom.xml
```

In the **src/main/ats** folder you can create ATS scripts using notepad with - *.ats* - extension.

## Sample ATS project

You can download an ATS test project : https://github.com/pierrehub2b/ats_test/archive/master.zip

Unzip this folder and now you can edit *.atsProjectProperties* and *pom.xml* files according to your needs. 

You can launch test of the project using java command line, Ant target or Maven goal.


## Execute project

Generate and compile ATS project using Maven and execute - *compile* - goal.
* Java files generated from ats scripts will be created into the *target/generated* folder of the project.
* Compiled classes will be created classes into the *target/classes* folder of the project.

```
cd [path-to-your-ats-project]
mvn compile
```

You can now use TestNG suite files to define your testing campaigns, please visit [TestNG](http://testng.org/doc/) to see how use TestNG with your favorite Java IDE or to create test suite executions.

**Or**

You can generate and compile project using command line with java (see 'Installing' chapter for ATS distribution path)

* Generate java files

```
cd [path-to-your-ats-project]
java -cp [ats-distribution-path]\libs/*;libs/* com.ats.generator.Generator
```
Java files will be generated into project folder *target/generated* from scripts found in *src/main/ats* folder. 
Java files in *src/main/java* folder will be copied into the *target/generated* folder.

* Generate and compile project (for this command using JRE is not enough, you need to use a JDK to compile classes)

```
cd [path-to-your-ats-project]
java -cp [ats-distribution-path]\libs/*;libs/* com.ats.generator.Generator -comp
```
Generated and copied classes from *target/generated* folder will be compiled into *target/classes* folder and all files.
Contents of folder *src/assets/* will be copied into the *target/classes* folder.

Tests are ready to be launched with TestNG but you have to first create and define a TestNG suite xml file, in this file you can define groups, package or scripts you want to include or exclude from execution.
Here the command line to launch tests defined in 'suite.xml' file :

```
cd [path-to-your-ats-project]
java -cp [ats-distribution-path]\libs/*;libs/*;target/classes org.testng.TestNG src/exec/suite.xml
```

**Or**

You can generate, compile and execute project using command line with Java and Maven (see 'Installing' chapter for ATS distribution path)

* Execute Maven

```
cd [path-to-your-ats-project]
mvn clean test
```

The default TestNG suite in the pom.xml file will be executed, this suite is define with *'maven-surefire-plugin'* and *'suiteXmlFile'* property.

## Customize ATS on host machine

Each machine running ATS scripts has it's own performance or available resources to execute automated tests.
You can change configuration on the installed ATS distribution like wait action by browsers or tested application installation path.

Here is an example of global ATS configuration file (*.atsProperties* file in ATS root install folder)

```
<?xml version="1.0" encoding="UTF-8"?>
<execute>
	<proxy>
		<type>system</type>
	</proxy>
	<appBounding>
		<x>20</x>
		<y>20</y>
		<width>1300</width>
		<height>960</height>
	</appBounding>
	<maxTry>
		<searchElement>15</searchElement>
		<getProperty>10</getProperty>
		<webService>1</webService>		
	</maxTry>
	<timeOut>
		<script>60</script>
		<pageLoad>120</pageLoad>
		<watchDog>180</watchDog>
		<webService>20</webService>
	</timeOut>
	<browsers>
		<browser>
			<name>chrome</name>
			<options>
				<option>--disable-infobars</option>
				<option>--disable-web-security</option>
			</options>
		</browser>
		<browser>
			<name>opera</name>
			<path>C:\Program Files\Opera\52.0.2871.40\opera.exe</path>
			<waitAction>150</waitAction>
		</browser>
		<browser>
			<name>firefox</name>
			<waitAction>200</waitAction>
			<waitProperty>70</waitProperty>
		</browser>
	</browsers>
	<applications>
		<application>
			<name>notepad++</name>
			<path>C:\Program Files (x86)\Notepad++\notepad++.exe</path>
		</application>
		<application>
			<name>filezilla</name>
			<path>C:\Program Files\FileZilla FTP Client\filezilla.exe</path>
		</application>
	</applications>
	<mobiles>
		<mobile>
			<name>settings</name>
			<endpoint>192.168.0.6:8080</endpoint>
			<package>settings</package>
			<waitAction>100</waitAction>
		</mobile>
	</mobiles>
	<apis>
		<api>
			<name>fakerest</name>
			<url>https://fakerestapi.azurewebsites.net</url>
		</api>
	</apis>
</execute>
```
* You can define the proxy used by browsers during execution of the test.
Proxy types available are *'system'*, *'auto'*, *'direct'* and *'manual'*, if *'manual'* type is selected you have to define host and port of the proxy :
```
<proxy>
	<type>manual</type>
	<host>proxyhost</host>
	<port>8080</port>
</proxy>
```
* **[appBounding]** : initial window size of a tested application and it's initial position
* **[maxTry -> searchElement]** : action's maxTry to wait element exists or wait element is interactable before execute
* **[maxTry -> getProperty]** : action's maxTry get property to wait element's property exists and property is not null
* **[maxTry -> webService]** : api REST or SOAP webservice max try if request fail with timeout
* **[timeOut -> script]** : javascript execution timeout (in sec.) for web application only
* **[timeOut -> pageLoad]** : page load timeout (in sec.) for web application only
* **[timeOut -> watchDog]** : action's execution timeout (in sec.) to prevent browser or application infinite hangup
* **[timeOut -> webService]** : api REST or SOAP webservice requests execution timeout (in sec.)
* For each browser used you can define following parameters :
  - **[path]** : the application binary file path
  - **[waitAction]** : wait after each actions (in millisec.)
  - **[waitProperty]** : wait for double check attributes value (in millisec.) 
* You can add options to the browsers driver
* You can define application name and path of installed applications on the host machine
* You can define a chromium like browser in the available browsers list. In order to use a chromium like browser you have to define *'name'*, *'driver'* and *'path'* attributes of the browser element. The *'driver'* attribute is the driver file *(without extension)* in the ATS drivers folder and the *'path'* attribute is the executable file of the chromium browser :
```
<browser>
	<name>chromium</name>
	<driver>chromiumdriver</driver>
	<path>C:\Program\chromium\chrome.exe</path>
</browser>
```
* If you want to use JxBrowser based application, you have to define *'name'*, *'driver'* and *'path'* attributes of the browser element. The *'driver'* attribute is the driver file *(without extension)* in the ATS drivers folder and the *'path'* attribute is the executable file of the JxBrowser start script, the remote debug port used by default is 9222 (you have to enable debug mode to the JxBrowser based application).
You can find all chromium driver versions here : https://sites.google.com/a/chromium.org/chromedriver, you have to choose the version corresponding to your version of chromium engine in your JxBrowser based application:
```
<browser>
	<name>jx</name>
	<driver>jxbrowser</driver>
	<path>C:\Program\path_to_jxbrowser_based_application\app.exe</path>
</browser>
```
*Chromium binaries files and Chromium driver for Windows can be found here : http://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64/650848/*
* ATS will automatically be updated with the last version of the supported browsers, in order to use another browser driver version you can define an executable driver name in the *'driver'* property of a named browser. If you want to use the version 73 of the chromedriver you have to download the right version of the driver, rename it, and copy it into the *'drivers'* folder :
```
<browser>
	<name>chrome</name>
	<driver>chromedriver_73</driver>
</browser>
```
* You can define a Neoload configuration to record and design load testing project with Neoload. ATS Neoload actions will enable Neoload UserPath recording and send commands to execute some design actions in a Neoload project. In order to enable Neoload recording you have to define Neoload proxy and service parameters :
```
<neoload>
	<recorder>
		<host>192.168.0.100</host>
		<port>8090</port>
	</recorder>
	<design>
		<api>Design/v1/Service.svc</api>
		<port>7400</port>
	</design>
</neoload>
```
## List of available ATS actions

```
channel-start : Start a new application channel with specified name
channel-switch : Switch to specified channel name (send application window to foreground)
channel-close : Close specified channel name
goto-url : Navigate to specified url
subscript : Call Ats or Java subscript
keyboard : Enter keyboard action
click : Simple click on found element
over : Make mouse over action on found element
drag : Mouse start drag action on found element
drop : Mouse drop action on found element
swipe : Execute mouse swipe gesture on found element
scroll : Execute mouse wheel action
check-property : Check property value of found element
check-count : Check number of element found with defined criterias
check-value : Make comparison between two values
property : Catch property value of found element into variable
window-resize : Resize or relocate current channel application window
window-switch : Switch between windows (or tabs) of current channel
window-close : Close indexed window
select : Select item by value, text or index of found select element
javascript : Execute javascript code on found element
comment : Add comment to script
```

## Thirdparty components

* [Neoload](https://www.neotys.com/) - Load testing platform
* [Selenium](https://www.seleniumhq.org/) - The testing framework used
* [TestNG](http://testng.org/doc/) - Test runner

## Authors

* **Pierre Huber** - *Initial work* - [Pierrehub2b](https://github.com/pierrehub2b)

See also the list of [contributors](https://github.com/pierrehub2b/actiontestscript/graphs/contributors) who participated in this project.

## License

This project is licensed under the Apache 2.0 License - see the [LICENSE.md](LICENSE.md) file for details

CHANGELOG
v1.5.5
==================

* Fix csv read file using UTF8 charset
* Fix stale reference exception on mouse click action
* Add JxBrowser support

src/main/java/com/ats/driver/AtsManager.java
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
 */

package com.ats.driver;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import com.ats.executor.TestBound;
import com.ats.generator.ATS;
import com.ats.tools.Utils;

public class AtsManager {

	public static final String ATS_FOLDER = ".actiontestscript";

	private static final String DRIVERS_FOLDER = "drivers";
	private static final String ATS_PROPERTIES_FILE = ".atsProperties";

	private static final Double APPLICATION_WIDTH = 1280.00;
	private static final Double APPLICATION_HEIGHT = 960.00;

	private static final Double APPLICATION_X = 10.00;
	private static final Double APPLICATION_Y = 10.00;

	private static final int SCRIPT_TIMEOUT = 60;
	private static final int PAGELOAD_TIMEOUT = 120;
	private static final int WATCHDOG_TIMEOUT = 200;
	private static final int WEBSERVICE_TIMEOUT = 20;

	private static final int MAX_TRY_SEARCH = 15;
	private static final int MAX_TRY_PROPERTY = 10;
	private static final int MAX_TRY_WEBSERVICE = 1;

	private static final int SCROLL_UNIT = 100;

	//-----------------------------------------------------------------------------------------------------
	//-----------------------------------------------------------------------------------------------------

	private Path driversFolderPath;
	private Properties properties;

	private double applicationWidth = APPLICATION_WIDTH;
	private double applicationHeight = APPLICATION_HEIGHT;

	private double applicationX = APPLICATION_X;
	private double applicationY = APPLICATION_Y;

	private int scrollUnit = SCROLL_UNIT;

	private int scriptTimeOut = SCRIPT_TIMEOUT;
	private int pageloadTimeOut = PAGELOAD_TIMEOUT;
	private int watchDogTimeOut = WATCHDOG_TIMEOUT;
	
	private int webServiceTimeOut = WEBSERVICE_TIMEOUT;

	private int maxTrySearch = MAX_TRY_SEARCH;
	private int maxTryProperty = MAX_TRY_PROPERTY;
	private int maxTryWebservice = MAX_TRY_WEBSERVICE;

	private AtsProxy proxy;
	private AtsProxy neoloadProxy;

	private String neoloadDesignApi;

	private List<ApplicationProperties> applicationsList = new ArrayList<ApplicationProperties>();

	public static String getVersion() {

		final InputStream resourceAsStream = AtsManager.class.getResourceAsStream("/version.properties");
		Properties prop = new Properties();
		try{
			prop.load( resourceAsStream );
			return prop.getProperty("version");
		}catch(Exception e) {}

		return null;
	}

	public AtsManager() {

		String atsHome = System.getProperty("ats.home");
		if(atsHome == null || atsHome.length() == 0) {
			atsHome = System.getenv("ATS_HOME");
			if(atsHome == null || atsHome.length() == 0) {
				atsHome = System.getProperty("user.home") + File.separator + ATS_FOLDER;
			}
		}

		final Path atsFolderPath = Paths.get(atsHome);
		if(atsFolderPath.toFile().exists()) {
			properties = loadProperties(atsFolderPath.resolve(ATS_PROPERTIES_FILE));
			driversFolderPath = atsFolderPath.resolve(DRIVERS_FOLDER);
		}else {
			ATS.logError("ATS folder not found -> " + atsHome);
			System.exit(0);
		}

		if(proxy == null) {
			proxy = new AtsProxy(AtsProxy.SYSTEM);
		}
	}

	private Properties loadProperties(Path propertiesPath) {

		final File xmlFile = propertiesPath.toFile();
		if(xmlFile.exists()) {
			try {

				final DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();

				try {
					final DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
					final Document doc = dBuilder.parse(xmlFile);

					doc.getDocumentElement().normalize();

					NodeList nl0 = doc.getElementsByTagName("browser");
					if(nl0 != null && nl0.getLength() > 0) {
						for (int temp = 0; temp < nl0.getLength(); temp++) {
							Node browser = nl0.item(temp);
							if (browser.getNodeType() == Node.ELEMENT_NODE) {
								Element browserElement = (Element) browser;
								if(browserElement.hasChildNodes() && browserElement.getChildNodes().getLength() > 1) {
									NodeList nodeList = browserElement.getElementsByTagName("name");
									if(nodeList != null && nodeList.getLength() > 0) {
										if(nodeList.item(0).getChildNodes().getLength() > 0) {
											String name = nodeList.item(0).getChildNodes().item(0).getNodeValue();
											String path = null;
											String wait = null;
											String check = null;
											String lang = null;
											String driver = null;

											nodeList = browserElement.getElementsByTagName("path");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													path = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											nodeList = browserElement.getElementsByTagName("driver");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													driver = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											nodeList = browserElement.getElementsByTagName("lang");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													lang = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											nodeList = browserElement.getElementsByTagName("waitAction");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													wait = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											nodeList = browserElement.getElementsByTagName("waitProperty");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													check = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}
											addApplicationProperties(ApplicationProperties.BROWSER_TYPE, name, driver, path, wait, check, lang);
										}
									}
								}
							}
						}
					}

					nl0 = doc.getElementsByTagName("appBounding");
					if(nl0 != null && nl0.getLength() > 0) {
						NodeList bound = ((Element)nl0.item(0)).getElementsByTagName("width");
						if(bound != null && bound.getLength() > 0) {
							try {
								applicationWidth = Double.parseDouble(bound.item(0).getChildNodes().item(0).getNodeValue());
							}catch(NumberFormatException e){}
						}

						bound = ((Element)nl0.item(0)).getElementsByTagName("height");
						if(bound != null && bound.getLength() > 0) {
							try {
								applicationHeight = Double.parseDouble(bound.item(0).getChildNodes().item(0).getNodeValue());
							}catch(NumberFormatException e){}
						}

						bound = ((Element)nl0.item(0)).getElementsByTagName("x");
						if(bound != null && bound.getLength() > 0) {
							try {
								applicationX = Double.parseDouble(bound.item(0).getChildNodes().item(0).getNodeValue());
							}catch(NumberFormatException e){}
						}

						bound = ((Element)nl0.item(0)).getElementsByTagName("y");
						if(bound != null && bound.getLength() > 0) {
							try {
								applicationY = Double.parseDouble(bound.item(0).getChildNodes().item(0).getNodeValue());
							}catch(NumberFormatException e){}
						}
					}

					nl0 = doc.getElementsByTagName("timeOut");
					if(nl0 != null && nl0.getLength() > 0) {
						NodeList timeOut = ((Element)nl0.item(0)).getElementsByTagName("script");
						if(timeOut != null && timeOut.getLength() > 0) {
							scriptTimeOut = Utils.string2Int(timeOut.item(0).getChildNodes().item(0).getNodeValue(), SCRIPT_TIMEOUT);
						}

						timeOut = ((Element)nl0.item(0)).getElementsByTagName("pageLoad");
						if(timeOut != null && timeOut.getLength() > 0) {
							pageloadTimeOut = Utils.string2Int(timeOut.item(0).getChildNodes().item(0).getNodeValue(), PAGELOAD_TIMEOUT);
						}

						timeOut = ((Element)nl0.item(0)).getElementsByTagName("watchDog");
						if(timeOut != null && timeOut.getLength() > 0) {
							watchDogTimeOut = Utils.string2Int(timeOut.item(0).getChildNodes().item(0).getNodeValue(), WATCHDOG_TIMEOUT); 
						}
						
						timeOut = ((Element)nl0.item(0)).getElementsByTagName("webService");
						if(timeOut != null && timeOut.getLength() > 0) {
							webServiceTimeOut = Utils.string2Int(timeOut.item(0).getChildNodes().item(0).getNodeValue(), WEBSERVICE_TIMEOUT);
						}
					}

					nl0 = doc.getElementsByTagName("maxTry");
					if(nl0 != null && nl0.getLength() > 0) {
						NodeList maxTryNode = ((Element)nl0.item(0)).getElementsByTagName("searchElement");
						if(maxTryNode != null && maxTryNode.getLength() > 0) {
							maxTrySearch = Utils.string2Int(maxTryNode.item(0).getChildNodes().item(0).getNodeValue(), MAX_TRY_SEARCH);
						}

						maxTryNode = ((Element)nl0.item(0)).getElementsByTagName("getProperty");
						if(maxTryNode != null && maxTryNode.getLength() > 0) {
							maxTryProperty = Utils.string2Int(maxTryNode.item(0).getChildNodes().item(0).getNodeValue(), MAX_TRY_PROPERTY);
						}
						
						maxTryNode = ((Element)nl0.item(0)).getElementsByTagName("webService");
						if(maxTryNode != null && maxTryNode.getLength() > 0) {
							maxTryWebservice = Utils.string2Int(maxTryNode.item(0).getChildNodes().item(0).getNodeValue(), MAX_TRY_WEBSERVICE);
							if(maxTryWebservice < 1) {
								maxTryWebservice = MAX_TRY_WEBSERVICE;
							}
						}
					}

					nl0 = doc.getElementsByTagName("proxy");
					if(nl0 != null && nl0.getLength() > 0) {
						NodeList data = ((Element)nl0.item(0)).getElementsByTagName("type");
						if(data != null && data.getLength() > 0) {

							String type = data.item(0).getChildNodes().item(0).getNodeValue();

							if(AtsProxy.DIRECT.equals(type)) {

								proxy = new AtsProxy(AtsProxy.DIRECT);

							}else if(AtsProxy.AUTO.equals(type)) {

								proxy = new AtsProxy(AtsProxy.AUTO);

							}else if(AtsProxy.MANUAL.equals(type)) {

								String host = null;
								int port = -1;

								data = ((Element)nl0.item(0)).getElementsByTagName("host");
								if(data != null && data.getLength() > 0) {
									host = data.item(0).getChildNodes().item(0).getNodeValue();
								}

								data = ((Element)nl0.item(0)).getElementsByTagName("port");
								if(data != null && data.getLength() > 0) {
									port = Utils.string2Int(data.item(0).getChildNodes().item(0).getNodeValue());
								}

								if(host != null && port > 0) {
									proxy = new AtsProxy(AtsProxy.MANUAL, host, port);
								}
							}
						}
					}

					nl0 = doc.getElementsByTagName("recorder");
					if(nl0 != null && nl0.getLength() > 0) {
						for (int temp = 0; temp < nl0.getLength(); temp++) {
							Node recorder = nl0.item(temp);
							if (recorder.getNodeType() == Node.ELEMENT_NODE) {
								Element recorderElement = (Element) recorder;
								if(recorderElement.hasChildNodes() && recorderElement.getChildNodes().getLength() > 1) {

									String host = null;
									NodeList nl = recorderElement.getElementsByTagName("host");
									if(nl != null && nl.getLength() > 0) {
										if(nl.item(0).getChildNodes().getLength() > 0) {
											host = nl.item(0).getChildNodes().item(0).getNodeValue();
										}
									}

									String port = null;
									nl = recorderElement.getElementsByTagName("port");
									if(nl != null && nl.getLength() > 0) {
										if(nl.item(0).getChildNodes().getLength() > 0) {
											port = nl.item(0).getChildNodes().item(0).getNodeValue();
										}
									}

									if(host != null && port != null) {
										neoloadProxy = new AtsProxy(host, port);
									}
								}
							}
						}
					}

					if(neoloadProxy != null) {
						nl0 = doc.getElementsByTagName("design");
						if(nl0 != null && nl0.getLength() > 0) {
							for (int temp = 0; temp < nl0.getLength(); temp++) {
								Node design = nl0.item(temp);
								if (design.getNodeType() == Node.ELEMENT_NODE) {
									Element designElement = (Element) design;
									if(designElement.hasChildNodes() && designElement.getChildNodes().getLength() > 1) {

										String api = null;
										NodeList nl = designElement.getElementsByTagName("api");
										if(nl != null && nl.getLength() > 0) {
											if(nl.item(0).getChildNodes().getLength() > 0) {
												api = nl.item(0).getChildNodes().item(0).getNodeValue();
												if(!api.startsWith("/")) {
													api = "/" + api;
												}
												if(!api.endsWith("/")) {
													api = api + "/";
												}
											}
										}

										int port = 0;
										nl = designElement.getElementsByTagName("port");
										if(nl != null && nl.getLength() > 0) {
											if(nl.item(0).getChildNodes().getLength() > 0) {
												port = Utils.string2Int(nl.item(0).getChildNodes().item(0).getNodeValue());
											}
										}

										if(api != null && port > 0) {
											neoloadDesignApi = "http://" + neoloadProxy.getHost() + ":" + port + api;
										}
									}
								}
							}
						}
					}

					nl0 = doc.getElementsByTagName("application");
					if(nl0 != null && nl0.getLength() > 0) {
						for (int temp = 0; temp < nl0.getLength(); temp++) {
							Node application = nl0.item(temp);
							if (application.getNodeType() == Node.ELEMENT_NODE) {
								Element applicationElement = (Element) application;
								if(applicationElement.hasChildNodes() && applicationElement.getChildNodes().getLength() > 1) {
									NodeList nodeList = applicationElement.getElementsByTagName("name");
									if(nodeList != null && nodeList.getLength() > 0) {
										if(nodeList.item(0).getChildNodes().getLength() > 0) {
											String name = nodeList.item(0).getChildNodes().item(0).getNodeValue();
											String path = "";
											String wait = null;

											nodeList = applicationElement.getElementsByTagName("path");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													path = nodeList.item(0).getChildNodes().item(0).getNodeValue();

													File checkFile = new File(path);
													if(!checkFile.exists() || !checkFile.isFile()) {
														path = "";
													}
												}
											}

											nodeList = applicationElement.getElementsByTagName("waitAction");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													wait = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											addApplicationProperties(ApplicationProperties.DESKTOP_TYPE, name, path, wait, "", null);
										}
									}
								}
							}
						}
					}

					nl0 = doc.getElementsByTagName("mobile");
					if(nl0 != null && nl0.getLength() > 0) {
						for (int temp = 0; temp < nl0.getLength(); temp++) {
							Node app = nl0.item(temp);
							if (app.getNodeType() == Node.ELEMENT_NODE) {
								Element mobileApp = (Element) app;
								if(mobileApp.hasChildNodes() && mobileApp.getChildNodes().getLength() > 1) {
									NodeList nodeList = mobileApp.getElementsByTagName("name");
									if(nodeList != null && nodeList.getLength() > 0) {
										if(nodeList.item(0).getChildNodes().getLength() > 0) {
											String name = nodeList.item(0).getChildNodes().item(0).getNodeValue();
											String url = "";
											String wait = null;

											nodeList = mobileApp.getElementsByTagName("endpoint");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													url = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											nodeList = mobileApp.getElementsByTagName("package");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													url += "/" + nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											nodeList = mobileApp.getElementsByTagName("waitAction");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													wait = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											addApplicationProperties(ApplicationProperties.MOBILE_TYPE, name, url, wait, "", null);
										}
									}
								}
							}
						}
					}

					nl0 = doc.getElementsByTagName("api");
					if(nl0 != null && nl0.getLength() > 0) {
						for (int temp = 0; temp < nl0.getLength(); temp++) {
							Node app = nl0.item(temp);
							if (app.getNodeType() == Node.ELEMENT_NODE) {
								Element mobileApp = (Element) app;
								if(mobileApp.hasChildNodes() && mobileApp.getChildNodes().getLength() > 1) {
									NodeList nodeList = mobileApp.getElementsByTagName("name");
									if(nodeList != null && nodeList.getLength() > 0) {
										if(nodeList.item(0).getChildNodes().getLength() > 0) {
											String name = nodeList.item(0).getChildNodes().item(0).getNodeValue();
											String url = "";
											String wait = null;

											nodeList = mobileApp.getElementsByTagName("url");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													url = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											nodeList = mobileApp.getElementsByTagName("waitAction");
											if(nodeList != null && nodeList.getLength() > 0) {
												if(nodeList.item(0).getChildNodes().getLength() > 0) {
													wait = nodeList.item(0).getChildNodes().item(0).getNodeValue();
												}
											}

											addApplicationProperties(ApplicationProperties.API_TYPE, name, url, wait, "", null);
										}
									}
								}
							}
						}
					}

				} catch (ParserConfigurationException e) {

				} catch (SAXException e) {

				}

			} catch (IOException e) {

			}
		}

		return new Properties();
	}

	private void addApplicationProperties(int type, String name, String path, String wait, String check, String lang) {
		addApplicationProperties(type, name, null, path, wait, check, lang);
	}

	private void addApplicationProperties(int type, String name, String driver, String path, String wait, String check, String lang) {
		applicationsList.add(new ApplicationProperties(type, name, driver, path, Utils.string2Int(wait, -1), Utils.string2Int(check, -1), lang));
	}

	//------------------------------------------------------------------------------------------------------------------
	// Getters
	//------------------------------------------------------------------------------------------------------------------

	public ApplicationProperties getApplicationProperties(String name) {
		for (int i=0; i < this.applicationsList.size(); i++) {
			final ApplicationProperties properties = this.applicationsList.get(i);
			if (name.equals(properties.getName())){
				return properties;
			}
		}
		return new ApplicationProperties(name);
	}

	public int getScriptTimeOut() {
		return scriptTimeOut;
	}

	public int getPageloadTimeOut() {
		return pageloadTimeOut;
	}

	public int getWatchDogTimeOut() {
		return watchDogTimeOut;
	}
	
	public int getWebServiceTimeOut() {
		return webServiceTimeOut;
	}

	public TestBound getApplicationBound() {
		return new TestBound(applicationX, applicationY, applicationWidth, applicationHeight);
	}

	public String getPropertyString(String key) {
		return properties.getProperty(key);
	}

	public Path getDriversFolderPath() {
		return driversFolderPath;
	}

	public int getMaxTrySearch() {
		return maxTrySearch;
	}

	public int getMaxTryProperty() {
		return maxTryProperty;
	}
	
	public int getMaxTryWebservice() {
		return maxTryWebservice;
	}

	public int getScrollUnit() {
		return scrollUnit;
	}
	
	public String getNeoloadDesignApi() {
		return neoloadDesignApi;
	}

	public AtsProxy getNeoloadProxy() {
		if(neoloadProxy != null) {
			return neoloadProxy;
		}
		return getProxy();
	}

	public AtsProxy getProxy() {
		return proxy;
	}
}